### PR TITLE
Infer parquet reader type based on file metadata (wip)

### PR DESCRIPTION
**Pull request by saurabhd336**
Instructions:
1. The PR has to be tagged with at least one of the following labels (*):
   1. `feature`
   2. `bugfix`
   3. `performance`
   4. `ui`
   5. `backward-incompat`
   6. `release-notes` (**)
2. Remove these instructions before publishing the PR.
 
(*) Other labels to consider:
- `testing`
- `dependencies`
- `docker`
- `kubernetes`
- `observability`
- `security`
- `code-style`
- `extension-point`
- `refactor`
- `cleanup`

(**) Use `release-notes` label for scenarios like:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed

### CODE REVIEW

**Comment by xiangfu0** [comment]
Can you add a sample data file with decimal field and test the file is correctly parsed.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordReader.java [hunk 1]
```diff
@@ -32,17 +38,34 @@
...
+        && ((ParquetRecordReaderConfig) recordReaderConfig).useParquetNativeRecordReader()) {
       _useAvroParquetRecordReader = false;
       _internalParquetRecordReader = new ParquetNativeRecordReader();
+    } else {
+      // No reader type specified. Determine using file metadata
+      ParquetMetadata parquetMetadata =
+          ParquetFileReader.readFooter(new Configuration(), new Path(dataFile.getAbsolutePath()),
+              ParquetMetadataConverter.NO_FILTER);
+      Map<String, String> fileKeyValueMeta = parquetMetadata.getFileMetaData().getKeyValueMetaData();
+      boolean useParquetAvroRecordReader = fileKeyValueMeta.containsKey(AVRO_SCHEMA_METADATA_KEY);
```
Comment (KKcorps) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordReader.java:61]: Might throw null pointer exception here if `fileKeyValueMeta` is null.

File: pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordReader.java [hunk 2]
```diff
@@ -32,17 +38,34 @@
...
+    } else if (recordReaderConfig != null
+        && ((ParquetRecordReaderConfig) recordReaderConfig).useParquetNativeRecordReader()) {
       _useAvroParquetRecordReader = false;
       _internalParquetRecordReader = new ParquetNativeRecordReader();
+    } else {
+      // No reader type specified. Determine using file metadata
+      ParquetMetadata parquetMetadata =
+          ParquetFileReader.readFooter(new Configuration(), new Path(dataFile.getAbsolutePath()),
+              ParquetMetadataConverter.NO_FILTER);
+      Map<String, String> fileKeyValueMeta = parquetMetadata.getFileMetaData().getKeyValueMetaData();
```
Comment (xiangfu0) [inline pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordReader.java:60]: You can put this check `hasAvroSchemaInParquetFile()` inside `org.apache.pinot.plugin.inputformat.parquet.ParquetUtils` and reuse the same method inside `org.apache.pinot.plugin.inputformat.parquet.ParquetUtils.getParquetAvroSchema(...)`

Comment (saurabhd336) [reply pinot-plugins/pinot-input-format/pinot-parquet/src/main/java/org/apache/pinot/plugin/inputformat/parquet/ParquetRecordReader.java]: Ack

**Review by xiangfu0 — APPROVED** [review]
lgtm

**Comment by Jackie-Jiang** [comment]
The test failure might be related:
```
2022-08-31T07:17:28.7248119Z [ERROR] Failures: 
2022-08-31T07:17:28.7249092Z [ERROR]   ParquetRecordReaderTest.testComparison:105->testComparison:125 expected [false] but found [true]
```

**Comment by saurabhd336** [comment]
> The test failure might be related:
> 
> ```
> 2022-08-31T07:17:28.7248119Z [ERROR] Failures: 
> 2022-08-31T07:17:28.7249092Z [ERROR]   ParquetRecordReaderTest.testComparison:105->testComparison:125 expected [false] but found [true]
> ```

@Jackie-Jiang ACK. Fixed the test and added a new test to validate file metadata based reader selection

**Comment by Jackie-Jiang** [comment]
Can you please modify the PR description to include the new config key for the record reader config? Also update the Pinot doc where applicable

**Pull request merged**